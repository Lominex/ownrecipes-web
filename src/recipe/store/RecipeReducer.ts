import ingredient from './IngredientReducer';
import subrecipes from './SubRecipeReducer';
import fq from '../utilts/formatQuantity';
import ReduxHelper from '../../common/store/ReduxHelper';
import { Recipe, RecipeAction, RecipeActionTypes, RecipesAction, RecipeState, RECIPE_STORE } from './types';

const defaultState: RecipeState = ReduxHelper.getItemReducerDefaultState<Recipe>(RECIPE_STORE);

const recipe = (state = defaultState, action: RecipesAction): RecipeState => {
  if (action.store !== RECIPE_STORE) return ReduxHelper.caseDefaultReducer(state, action, defaultState);

  const recipeAction = action as RecipeAction;
  switch (recipeAction.type) {
    case RecipeActionTypes.RECIPE_LOAD:
      {
        const actionRecipe = recipeAction.data;

        let subRecipes, ingredients;
        /*
        const hasRecipe = state.items?.find(t => t.id === actionRecipe.id) != null;
        if (state.items && hasRecipe) {
          return state.items.map(rec => {
            if (rec.id === recipeAction.data.id) {
              subRecipes = subrecipes(
                rec.subrecipes,
                { subrecipes: actionRecipe.subrecipes,
                  formatQuantity: fq.bind(this, rec.servings, actionRecipe.servings),
                  type: action.type }
              );
              ingredients = ingredient(
                rec.ingredient_groups,
                { ingredient_groups: actionRecipe.ingredient_groups,
                  formatQuantity: fq.bind(this, rec.servings, actionRecipe.servings),
                  type: action.type }
              );
              return {
                ...actionRecipe,
                subrecipes: subRecipes,
                ingredient_groups: ingredients,
              };
            }
            return rec;
          });
        } else {
          subRecipes = subrecipes(
            [],
            { subrecipes: actionRecipe.subrecipes,
              formatQuantity: fq.bind(this, actionRecipe.servings, actionRecipe.servings),
              type: action.type }
          );
          ingredients = ingredient(
            [],
            { ingredient_groups: actionRecipe.ingredient_groups,
              formatQuantity: fq.bind(this, actionRecipe.servings, actionRecipe.servings),
              type: action.type }
          );
          */

          const updItem = {
            ...actionRecipe,
            subrecipes: subRecipes ?? [],
            ingredient_groups: ingredients ?? [],
            customServings: actionRecipe.servings,
          };

          return ReduxHelper.setItem(state, updItem);
        // }
      }
    case RecipeActionTypes.RECIPE_DELETE:
      return defaultState;
    case RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE:
      /*
      return state.items?.map(recipe => {
        if (recipe.slug === action.recipeSlug) {
          action.servings = recipe.servings;
          const subRecipes = subrecipes(
            recipe.subrecipes,
            { formatQuantity: fq.bind(this, recipe.servings, action.customServings),
              type: action.type }
          );
          const ingredients = ingredient(
            recipe.ingredient_groups,
            { formatQuantity: fq.bind(this, recipe.servings, action.customServings),
              type: action.type }
          );
          return {
            ...recipe,
            subrecipes: subRecipes,
            ingredient_groups: ingredients,
            customServings: action.customServings,
          };
        }
        return recipe;
      }); */
      return state;
    /*
    case RecipeActionTypes.RECIPE_INGREDIENT:
    case RecipeActionTypes.RECIPE_INGREDIENT_CHECK_INGREDIENT:
    case RecipeActionTypes.RECIPE_INGREDIENT_CHECK_SUBRECIPE:
    case RecipeActionTypes.RECIPE_INGREDIENT_CHECK_ALL:
    case RecipeActionTypes.RECIPE_INGREDIENT_UNCHECK_ALL:
    case RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_RESET:
      return state.items?.map(recipe => {
        if (recipe.slug === action.recipeSlug) {
          return {
            ...recipe,
            subrecipes: subrecipes(recipe.subrecipes, action),
            ingredient_groups: ingredient(recipe.ingredient_groups, action),
          };
        }
        return recipe;
      }); */
      return state;
    default: return ReduxHelper.caseDefaultReducer(state, action, defaultState);
  }
};

export default recipe;
