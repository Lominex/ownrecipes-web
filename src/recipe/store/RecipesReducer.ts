import ReduxHelper from '../../common/store/ReduxHelper';
import { Recipe, RecipeAction, RecipeActionTypes, RecipesAction, RecipesState, RECIPES_STORE, RECIPE_STORE } from './types';

const defaultState: RecipesState = ReduxHelper.getMapReducerDefaultState<Recipe>(RECIPES_STORE);

const recipes = (state = defaultState, action: RecipesAction): RecipesState => {
  if (action.store === RECIPE_STORE) {
    const recipeAction = action as RecipeAction;
    switch (recipeAction.type) {
      case RecipeActionTypes.RECIPE_LOAD:
        return ReduxHelper.setMapItem(state, recipeAction.data.slug, recipeAction.data);
      case RecipeActionTypes.RECIPE_DELETE:
        return ReduxHelper.deleteMapItem(state, recipeAction.data);
      default: break;
    }
  }

  return ReduxHelper.caseDefaultReducer(state, action, defaultState);
};

export default recipes;
