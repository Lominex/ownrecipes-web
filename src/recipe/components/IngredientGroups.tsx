import { defineMessages, useIntl } from 'react-intl';
import Spinner from 'react-spinkit';
import P from '../../common/components/P';
import { PendingState } from '../../common/store/GenericReducerType';
import { IngredientGroup } from '../store/types';

import Ingredients from './Ingredients';

export interface IIngredientGroupsProps {
  groups: Array<IngredientGroup>;
  pending: keyof typeof PendingState;
  checkIngredient: (id: number, checked: boolean) => void;
}

const IngredientGroups: React.FC<IIngredientGroupsProps> = ({ groups, pending, checkIngredient }: IIngredientGroupsProps) => {
  const { formatMessage } = useIntl();

  const messages = defineMessages({
    no_ingredients: {
      id: 'recipe.recipe_ingredient_button.no_ingredients',
      description: 'No ingredients provided message',
      defaultMessage: '(This recipe has no ingredients.)',
    },
  });

  const ingredientGroups = groups.map(group => (
    <div className='ingredient-group' key={group.title}>
      {group.title && <b>{group.title}</b>}
      <Ingredients data={group.ingredients} checkIngredient={checkIngredient} />
    </div>
  ));

  return (
    <div className='ingredient-groups'>
      {pending === PendingState.LOADING && <Spinner name='circle' className='recipe-list-spinner' fadeIn='none' />}
      {pending === PendingState.COMPLETED && ingredientGroups.length === 0 && (
        <P className='placeholder'>{formatMessage(messages.no_ingredients)}</P>
      )}
      {ingredientGroups}
    </div>
  );
};

export default IngredientGroups;
