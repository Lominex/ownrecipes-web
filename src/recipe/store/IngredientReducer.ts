import { Dispatch as ReduxDispatch } from 'redux';
import { BasicAction, PayloadAction } from '../../common/store/ReduxHelper';
import { Ingredient, IngredientGroup } from './RecipeTypes';

export const RECIPE_INGREDIENTS_STORE = '@@recipeIngredients';

export type QuantityFormatter = (numerator: number | undefined, denominator: number) => string;

export enum RecipeIngredientReducerActionTypes {
  RECIPE_INGREDIENTS_LOAD = 'RECIPE_INGREDIENTS_LOAD',
  RECIPE_INGREDIENTS_SERVINGS_UPDATE = 'RECIPE_INGREDIENTS_SERVINGS_UPDATE',
}

export type IRecipeIngredientLoadAction = {
  store: typeof RECIPE_INGREDIENTS_STORE;
  typs: typeof RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_LOAD;
  formatQuantity: (numerator: number | undefined, denominator: number) => string;
} & PayloadAction<Array<IngredientGroup>>;

export type IRecipeIngredientServingsUpdateAction = {
  store: typeof RECIPE_INGREDIENTS_STORE;
  typs: typeof RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_SERVINGS_UPDATE;
  formatQuantity: (numerator: number | undefined, denominator: number) => string;
} & BasicAction;

export type RecipeIngredientsState = Array<IngredientGroup>;
export type RecipeIngredientsAction = IRecipeIngredientLoadAction | IRecipeIngredientServingsUpdateAction;
export type RecipeIngredientsDispatch  = ReduxDispatch<RecipeIngredientsAction>;

type IngredientReduceFunction = (ingr: Ingredient) => Ingredient;

const ingredients = (state: RecipeIngredientsState, cb: IngredientReduceFunction): RecipeIngredientsState => state.map(ig => ({
  ...ig,
  ingredients: ig.ingredients.map(ingredient => cb(ingredient)),
}));

const merge = (state: RecipeIngredientsState, ingredientGroups: Array<IngredientGroup>, formatQuantity: QuantityFormatter) => {
  const list: Array<number> = [];
  state
    .map(ig => ig.ingredients)
    .forEach(ingrArray => {
      ingrArray
        .filter(ingr => ingr.checked)
        .forEach(ingr => list.push(ingr.id));
    });

  return ingredients(ingredientGroups, ingr => {
    const checked = list.includes(ingr.id);
    const custom = formatQuantity(ingr.numerator, ingr.denominator);
    return { ...ingr, quantity: custom, checked: checked };
  });
};

const defaultState: RecipeIngredientsState = [];

const recipes = (state = defaultState, action: RecipeIngredientsAction): RecipeIngredientsState => {
  if (action.store === RECIPE_INGREDIENTS_STORE) {
    switch (action.typs) {
      case RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_LOAD:
        return merge(state, action.payload, action.formatQuantity);
      case RecipeIngredientReducerActionTypes.RECIPE_INGREDIENTS_SERVINGS_UPDATE:
        return ingredients(state, i => {
          const custom = action.formatQuantity(i.numerator, i.denominator);
          return { ...i, quantity: custom };
        });
      /*
      case RecipeActionTypes.RECIPE_INGREDIENT_CHECK_INGREDIENT:
        return ingredients(state, i => {
          if (i.id === action.id) {
            return { ...i, checked: action.value };
          }
          return i;
        });
      case RecipeActionTypes.RECIPE_INGREDIENT_CHECK_ALL:
        return ingredients(state, i => ({ ...i, checked: true }));
      case RecipeActionTypes.RECIPE_INGREDIENT_UNCHECK_ALL:
        return ingredients(state, i => ({ ...i, checked: false }));
      */
      default:
        break;
    }
  }

  return state;
};

export default recipes;
