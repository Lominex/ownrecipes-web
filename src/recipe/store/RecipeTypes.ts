import { Dispatch as ReduxDispatch } from 'redux';

import { NUMBER_UNDEFINED } from '../../common/constants';
import ItemReducerType from '../../common/store/ItemReducerType';
import { GenericItemReducerAction } from '../../common/store/ReduxHelper';

export type Quantity = {
  numerator?:   number;
  denominator:  number;
  measurement?: string;
}

export type IngredientInput = {
  title:        string;

  quantity?:    string;
  checked?:     boolean;
} & Partial<Quantity>;

export type IngredientDto = {
  id:           number;
  title:        string;
} & Quantity;
export type Ingredient = {
  id:           number;
  title:        string;

  quantity?:    string;
  checked?:     boolean;
} & Quantity;
export const toIngredientDto = (obj: Ingredient): IngredientDto => ({
  id:    obj.id,
  title: obj.title,

  numerator:   obj.numerator,
  denominator: obj.denominator,
  measurement: obj.measurement,
});
export const toIngredient = (dto: IngredientDto): Ingredient => ({
  id:    dto.id,
  title: dto.title,

  numerator:   dto.numerator && dto.numerator > 0 ? dto.numerator : undefined,
  denominator: dto.denominator,
  measurement: dto.measurement,
});

export type IngredientGroupDto = {
  title:       string;
  ingredients: Array<IngredientDto>;
}
export type IngredientGroup = {
  title:       string;
  ingredients: Array<Ingredient>;
}
export const toIngredientGroupDto = (obj: IngredientGroup): IngredientGroupDto => ({
  title:       obj.title,
  ingredients: obj.ingredients.map(i => toIngredientDto(i)),
});
export const toIngredientGroup = (dto: IngredientGroupDto): IngredientGroup => ({
  title:       dto.title,
  ingredients: dto.ingredients.map(i => toIngredient(i)),
});

export type SubRecipeDto = {
  title:       string;
  slug:        string;
  child_recipe_id: number;
  parent_recipe_id?: number;

  measurement?: string;
} & Quantity;

export type SubRecipe = {
  title:       string;
  slug:        string;
  child_recipe_id:   number;
  parent_recipe_id?: number;
  measurement?: string;

  quantity?:   string;
  checked?:    boolean;
} & Quantity;
export const toSubRecipeDto = (obj: SubRecipe): SubRecipeDto => ({
  title: obj.title,
  slug:  obj.slug,
  child_recipe_id:  obj.child_recipe_id,
  parent_recipe_id: obj.parent_recipe_id,

  measurement: obj.measurement,
  numerator:   obj.numerator,
  denominator: obj.denominator,
});
export const toSubRecipe = (dto: SubRecipeDto): SubRecipe => ({
  title: dto.title,
  slug:  dto.slug,
  child_recipe_id:  dto.child_recipe_id,
  parent_recipe_id: dto.parent_recipe_id,

  measurement: dto.measurement,
  numerator:   dto.numerator,
  denominator: dto.denominator,
});

export type CourseDto = {
  id:    number;
  title: string;
}

export type Course = {
  id:    number;
  title: string;
}

export const toCourse = (dto: CourseDto): Course => ({
  id:    dto.id,
  title: dto.title,
});

export const toCourseDto = (obj: Course): CourseDto => ({
  id:    obj.id,
  title: obj.title,
});

export type CuisineDto = {
  id:    number;
  title: string;
}

export type Cuisine = {
  id:    number;
  title: string;
}

export const toCuisine = (dto: CuisineDto): Cuisine => ({
  id:    dto.id,
  title: dto.title,
});

export const toCuisineDto = (obj: Cuisine): CuisineDto => ({
  id:    obj.id,
  title: obj.title,
});

export type TagDto = {
  id:    number;
  title: string;
  slug:  string;
}

export type Tag = {
  id:    number;
  title: string;
  slug:  string;
}

export const toTag = (dto: TagDto): Tag => ({
  id:    dto.id,
  title: dto.title,
  slug:  dto.slug,
});

export type RecipeListDto = {
  id:    number;
  title: string; // Tasty Chili 24
  slug:  string; // tasty-werwerchili-4

  photo_thumbnail?: string | null;

  info:     string;
  rating:   number;
  pub_date: string; // 2011-05-20
}

export type RecipeList = {
  id:    number;
  title: string; // Tasty Chili 24
  slug:  string; // tasty-werwerchili-4

  photoThumbnail?: string;

  info:     string;
  rating:   number;
  pubDate:  Date; // 2011-05-20
}

export const toRecipeList = (dto: RecipeListDto): RecipeList => ({
  id:    dto.id,
  title: dto.title,
  slug:  dto.slug,

  photoThumbnail: dto.photo_thumbnail ?? undefined,

  info: dto.info,
  rating: dto.rating,
  pubDate: new Date(dto.pub_date),
});

export type RecipeDto = {
  username:    string;
  author:      number;
  source:      string;

  cook_time: number;
  prep_time: number;
  servings:  number;

  course:  Course;
  cuisine: Cuisine;
  tags:    Array<TagDto>;

  photo?: string | null;

  subrecipes: Array<SubRecipeDto>;
  ingredient_groups: Array<IngredientGroupDto>;

  directions: string;

  public: boolean;
  update_date: string;

  customServings: number;
} & RecipeListDto;

export type Recipe = {
  username:    string;
  author:      number;
  source:      string;

  cookTime?: number;
  prepTime?: number;
  servings:  number;

  course:  Course;
  cuisine: Cuisine;
  tags:    Array<Tag>;

  photo?: string;

  subrecipes: Array<SubRecipe>;
  ingredientGroups: Array<IngredientGroup>;

  directions: string;

  public: boolean;
  updateDate: Date;

  customServings: number;
} & RecipeList;

export const toRecipe = (dto: RecipeDto): Recipe => ({
  id:    dto.id,
  title: dto.title,
  slug:  dto.slug,

  username: dto.username,
  author:   dto.author,
  source:   dto.source,

  cookTime: parseBackendNumber(dto.cook_time),
  prepTime: parseBackendNumber(dto.prep_time),
  servings: dto.servings,

  course:  toCourse(dto.course),
  cuisine: toCuisine(dto.cuisine),
  tags:    dto.tags.map(t => toTag(t)),

  photo: dto.photo ?? undefined,
  photoThumbnail: dto.photo_thumbnail ?? undefined,

  subrecipes: dto.subrecipes.map(sr => toSubRecipe(sr)),
  ingredientGroups: dto.ingredient_groups.filter(ig => ig.ingredients.length > 0).map(ig => toIngredientGroup(ig)),

  directions: dto.directions,
  info: dto.info,

  rating: dto.rating,
  public: dto.public,
  pubDate: new Date(dto.pub_date),
  updateDate: new Date(dto.update_date),

  customServings: dto.servings,
});

export type RecipeRequest = {
  title:      string;

  source:     string;

  cook_time:  number;
  prep_time:  number;
  servings:   number;

  tags:       Array<TagDto>;
  course:     CourseDto;
  cuisine:    CuisineDto;

  subrecipes: Array<SubRecipeDto>;
  ingredient_groups: Array<IngredientGroupDto>;
  directions: string;
  info:       string;

  public:     boolean;
};

function ifNull<T>(val: T | undefined, d: T): T {
  if (val == null || (typeof val === 'string' && val === '')) return d;
  return val;
}
function parseBackendNumber(val: number): number | undefined {
  if (val === NUMBER_UNDEFINED) return undefined;
  return val;
}

export const toRecipeRequest = (obj: Recipe): RecipeRequest => ({
  title:      obj.title,

  source:     obj.source,

  cook_time:  ifNull(obj.cookTime, NUMBER_UNDEFINED),
  prep_time:  ifNull(obj.prepTime, NUMBER_UNDEFINED),
  servings:   obj.servings,

  tags:       obj.tags,
  course:     toCourseDto(obj.course),
  cuisine:    toCuisineDto(obj.cuisine),

  subrecipes: obj.subrecipes?.map(sr => toSubRecipeDto(sr)) ?? [],
  ingredient_groups: obj.ingredientGroups?.filter(ig => ig.ingredients.length > 0).map(ig => toIngredientGroupDto(ig)),
  directions: obj.directions,
  info:       obj.info,

  public:     obj.public,
});

export enum RecipeActionTypes {
  RECIPE_DELETE = 'RECIPE_DELETE',
  RECIPE_INGREDIENT = 'RECIPE_INGREDIENT',
  RECIPE_INGREDIENT_CHECK_INGREDIENT = 'RECIPE_INGREDIENT_CHECK_INGREDIENT',
  RECIPE_INGREDIENT_CHECK_SUBRECIPE = 'RECIPE_INGREDIENT_CHECK_SUBRECIPE',
  RECIPE_INGREDIENT_CHECK_ALL = 'RECIPE_INGREDIENT_CHECK_ALL',
  RECIPE_INGREDIENT_UNCHECK_ALL = 'RECIPE_INGREDIENT_UNCHECK_ALL',
  RECIPE_INGREDIENT_SERVINGS_UPDATE = 'RECIPE_INGREDIENT_SERVINGS_UPDATE',
  RECIPE_INGREDIENT_SERVINGS_RESET = 'RECIPE_INGREDIENT_SERVINGS_RESET',
  RECIPE_LIST_BLANK = 'RECIPE_LIST_BLANK',
  RECIPE_LIST_LOADING = 'RECIPE_LIST_LOADING',
  RECIPE_LIST_COMPLETE = 'RECIPE_LIST_COMPLETE',
  RECIPE_LIST_ERROR = 'RECIPE_LIST_ERROR',
}

export const RECIPE_STORE = '@@recipe';

export interface IRecipeSlugAction {
  store: typeof RECIPE_STORE;
  type:  typeof RecipeActionTypes.RECIPE_DELETE;
  data:  { slug: string };
}

export interface IRecipeIngredientUpdateServingAction {
  store: typeof RECIPE_STORE;
  type:  typeof RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE;
  recipeSlug:     string;
  customServings: number;
}

export interface IRecipeIngredientServingSimpleAction {
  store: typeof RECIPE_STORE;
  type: typeof RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_RESET
      | typeof RecipeActionTypes.RECIPE_INGREDIENT_CHECK_ALL
      | typeof RecipeActionTypes.RECIPE_INGREDIENT_UNCHECK_ALL;
  recipeSlug: string;
}

export interface IRecipeIngredientCheckIngredientAction {
  store: typeof RECIPE_STORE;
  type:  typeof RecipeActionTypes.RECIPE_INGREDIENT_CHECK_INGREDIENT;
  recipeSlug: string;
  id:    number;
  value: boolean;
}

export interface IRecipeIngredientCheckSubRecipeAction {
  store: typeof RECIPE_STORE;
  type:  typeof RecipeActionTypes.RECIPE_INGREDIENT_CHECK_SUBRECIPE;
  recipeSlug: string;
  id:    number;
  value: boolean;
}

export type RecipeState     = ItemReducerType<Recipe>;
export type RecipeAction    = IRecipeSlugAction | IRecipeIngredientUpdateServingAction
   | IRecipeIngredientServingSimpleAction | IRecipeIngredientCheckIngredientAction | IRecipeIngredientCheckSubRecipeAction | GenericItemReducerAction<Recipe>;
export type RecipeDispatch  = ReduxDispatch<RecipeAction>;
