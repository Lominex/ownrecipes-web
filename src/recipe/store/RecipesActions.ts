import { RecipeDispatch, RECIPES_STORE } from './types';
import { ACTION } from '../../common/store/ReduxHelper';

// eslint-disable-next-line import/prefer-default-export
export const softReset = () => (dispatch: RecipeDispatch) => {
  dispatch({ store: RECIPES_STORE, type: ACTION.SOFT_RESET });
};
