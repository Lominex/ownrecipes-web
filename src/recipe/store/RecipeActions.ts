import { handleError, request } from '../../common/CustomSuperagent';
import { serverURLs } from '../../common/config';
import { RecipeActionTypes, RecipeDispatch, RECIPE_STORE, toRecipe } from './RecipeTypes';
import { ACTION } from '../../common/store/ReduxHelper';

export const load = (recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({ store: RECIPE_STORE, type: ACTION.GET_START });
  request()
    .get(`${serverURLs.recipe}${recipeSlug}/`)
    .then(res => {
      dispatch({ store: RECIPE_STORE, type: ACTION.GET_SUCCESS, data: toRecipe(res.body) });
    })
    .catch(err => dispatch(handleError(err, RECIPE_STORE)));
};

export const deleteRecipe = (recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({ store: RECIPE_STORE, type: ACTION.DELETE_START });
  request()
    .get(`${serverURLs.recipe}${recipeSlug}/`)
    .then(() => {
      dispatch({ store: RECIPE_STORE, type: RecipeActionTypes.RECIPE_DELETE, data: recipeSlug });
    })
    .catch(err => dispatch(handleError(err, RECIPE_STORE)));
};

export const updateServings = (recipeSlug: string, value: number) => (dispatch: RecipeDispatch) => {
  dispatch({
    store: RECIPE_STORE,
    type: RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_UPDATE,
    recipeSlug: recipeSlug,
    customServings: value,
  });
};

export const resetServings = (recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({
    store: RECIPE_STORE,
    type: RecipeActionTypes.RECIPE_INGREDIENT_SERVINGS_RESET,
    recipeSlug: recipeSlug,
  });
};

export const checkIngredient = (recipeSlug: string, id: number, value: boolean) => (dispatch: RecipeDispatch) => {
  dispatch({
    store: RECIPE_STORE,
    type: RecipeActionTypes.RECIPE_INGREDIENT_CHECK_INGREDIENT,
    recipeSlug: recipeSlug,
    id: id,
    value: value,
  });
};

export const checkSubRecipe = (recipeSlug: string, id: number, value: boolean) => (dispatch: RecipeDispatch) => {
  dispatch({
    store: RECIPE_STORE,
    type: RecipeActionTypes.RECIPE_INGREDIENT_CHECK_SUBRECIPE,
    recipeSlug: recipeSlug,
    id: id,
    value: value,
  });
};

export const checkAll = (recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({
    store: RECIPE_STORE,
    type: RecipeActionTypes.RECIPE_INGREDIENT_CHECK_ALL,
    recipeSlug: recipeSlug,
  });
};

export const unCheckAll = (recipeSlug: string) => (dispatch: RecipeDispatch) => {
  dispatch({
    store: RECIPE_STORE,
    type: RecipeActionTypes.RECIPE_INGREDIENT_UNCHECK_ALL,
    recipeSlug: recipeSlug,
  });
};

/*
export const bulkAdd = (recipeState: Recipe, list: string) => (dispatch: RecipeDispatch) => {
  const format = (i: Ingredient) => {
    const quantity = i.quantity && i.quantity !== '0' ? `${i.quantity} ` : '';
    const measurement = i.measurement ? `${i.measurement} ` : '';
    return quantity + measurement + i.title;
  };

  let checkedIngredients = recipeState.ingredient_groups.map(item => item.ingredients.reduce((myList, ingredient) => {
      if (ingredient && ingredient.checked) {
        myList.push({ list: list, title: format(ingredient) });
      }
      return myList;
    }, []));

  const checkedSubRecipe = recipeState.subrecipes.reduce((myList, ingredient) => {
    if (ingredient && ingredient.checked) {
      myList.push({ list: list, title: format(ingredient) });
    }
    return myList;
  }, []);

  checkedIngredients = checkedIngredients.reduce((a, b) => a.concat(b), []).concat(checkedSubRecipe);

  if (checkedIngredients.length > 0) {
    dispatch({ type: RecipeConstants.RECIPE_LIST_LOADING });
    request()
      .post(serverURLs.bulk_list_item)
      .send(checkedIngredients)
      .then(res => {
        dispatch({ type: RecipeConstants.RECIPE_LIST_COMPLETE });
        dispatch({
          type: RecipeConstants.RECIPE_INGREDIENT_UNCHECK_ALL,
          recipeSlug: recipeState.slug,
        });
      })
      .catch(err => { dispatch({ type: RecipeConstants.RECIPE_LIST_ERROR }); });
  }
};
*/
