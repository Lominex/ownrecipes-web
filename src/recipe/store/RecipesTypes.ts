import { Dispatch as ReduxDispatch } from 'redux';

import MapReducerType from '../../common/store/MapReducerType';
import { GenericReducerAction } from '../../common/store/ReduxHelper';
import { Recipe } from './RecipeTypes';

export const RECIPES_STORE = '@@recipes';

export type RecipesState    = MapReducerType<Recipe>;
export type RecipesAction   = GenericReducerAction;
export type RecipesDispatch = ReduxDispatch<RecipesAction>;
