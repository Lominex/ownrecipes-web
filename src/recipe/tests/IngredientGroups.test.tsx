import IngredientGroups from '../components/IngredientGroups';
import createComponentWithIntlAndRouter from '../../test/createComponentWithIntlAndRouter';

import data from './data';
import { IngredientGroup } from '../store/RecipeTypes';

test('Ingredient Group component test', () => {
  // const checkIngredient = jest.fn();
  const component = createComponentWithIntlAndRouter(
    <IngredientGroups groups={data.ingredient_groups as unknown as Array<IngredientGroup>} /* checkIngredient={checkIngredient} */ />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
