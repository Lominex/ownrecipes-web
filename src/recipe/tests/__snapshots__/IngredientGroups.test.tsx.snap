// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingredient Group component test 1`] = `
Array [
  <div
    className="ingredient-group"
  >
    <h3
      className="subheading"
    >
      Chili
    </h3>
    <table
      className="table ingredients-table table table-sm table-striped"
    >
      <thead
        className="hideme"
      >
        <tr>
          <th>
            <span>
              Quantity
            </span>
          </th>
          <th>
            <span>
              Ingredient
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="ingredient"
        >
          <td
            className="quantity"
          >
            <span>
              1
               
              dash
            </span>
          </td>
          <td
            className="ingredient"
          >
            <span>
              black pepper
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="ingredient-group"
  >
    <h3
      className="subheading"
    >
      Spices
    </h3>
    <table
      className="table ingredients-table table table-sm table-striped"
    >
      <thead
        className="hideme"
      >
        <tr>
          <th>
            <span>
              Quantity
            </span>
          </th>
          <th>
            <span>
              Ingredient
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="ingredient"
        >
          <td
            className="quantity"
          >
            <span>
              1
               
              dash
            </span>
          </td>
          <td
            className="ingredient"
          >
            <span>
              black pepper
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="ingredient-group"
  >
    
    <table
      className="table ingredients-table table table-sm table-striped"
    >
      <thead
        className="hideme"
      >
        <tr>
          <th>
            <span>
              Quantity
            </span>
          </th>
          <th>
            <span>
              Ingredient
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="ingredient"
        >
          <td
            className="quantity"
          >
            <span>
              1
               
              dash
            </span>
          </td>
          <td
            className="ingredient"
          >
            <span>
              black pepper
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
